File: fletcher16.pas
unit fletcher16;
//Fletcher-16
//Author: domasz
//Version: 0.1 (2022-11-17)
//Licence: MIT

interface

uses SysUtils, HasherBase, Dialogs;

type THasherFletcher16 = class(THasherbase)
  private
    FHash, FHash2: Word;
  public
    constructor Create; override;
    procedure Update(Msg: PByte; Length: Integer); override;
    function Final: String; override;
end;

implementation

constructor THasherFletcher16.Create;
begin
  inherited Create;
  Check := '1EDE';
  FHash := 0;
  FHash2 := 0;
end;

procedure THasherFletcher16.Update(Msg: PByte; Length: Integer);
var i: Integer;
begin
  for i:=0 to Length-1 do begin
    FHash := (FHash + Msg^) mod 255;
    FHash2 := (FHash2 + FHash) mod 255;

    Inc(Msg);
  end;   
end;

function THasherFletcher16.Final: String;
begin
  FHash := (FHash2 shl 8) or FHash;

  Result := IntToHex(FHash, 4);
end;

initialization
  HasherList.RegisterHasher('Fletcher-16', THasherFletcher16);

end.

File: PHP/Algos/fletcher16.php
<?php
//Fletcher-32
//Author: domasz
//Version: 0.1 (2022-11-19)
//Licence: MIT

require_once 'HasherBase.php';

class HasherFletcher16 extends HasherBase
{
	protected $FHash, $FHash2 = 0;

	public function __construct() 
	{
		$this->Check = '1EDE';
		$this->FHash = 0;
		$this->FHash2 = 0;
	}

	public function Update($Msg, $Length)
	{
		for ($i=0; $i<$Length; $i++)
		{
			$this->FHash = ($this->FHash + ord($Msg[$i]) ) % 255;
			$this->FHash2 = ($this->FHash2 + $this->FHash) % 255;
		}
	}

	public function Finish()
	{
		$this->FHash = ($this->FHash2 << 8) | $this->FHash;
		return sprintf('%04X', $this->FHash);
	}
}

$HasherList->RegisterHasher('Fletcher-16', 'HasherFletcher16');